### ISSUE TITLE
No way to close inline reply with keyboard shortcuts

### ISSUE THREAD

**Issue by charlag** [nolanlawson/pinafore#915]
Hi!
Thanks for the awesome work with keyboard shortcuts.

I noticed a minor thing that you cannot close inline replies. You also cannot really unfocus reply with keyboard (you can do this with tabbing but maybe it would be more strainghtforward to close reply with escape key or something like that?).

Thanks!

**Comment by charlag** [comment]
Another issue with that is that the field is auto-focus when you go back from another page and you lose you scroll position

**Comment by charlag** [comment]
Oh, I'm sorry, I didn't realize that you can close it by clicking "reply" again. Icon colors are... non-obvious.
Still with keyboard it's tricky?

**Comment by nolanlawson** [comment]
Hmmm, it's tricky because if you tap <kbd>r</kbd> then you are in reply mode, meaning it's focused on the text field. So typing anything again should not be a keyboard shortcut; it should type into the textarea.

As for the scroll position being messed up, that's already captured in #840

I'm not sure I see a good solution to this problem. Yes you're right the icon selected states are non-obvious in some themes; which theme are you using?

**Comment by charlag** [comment]
I'm using Hacker.

Shift+tab, r works. Maybe that's enough?
I wonder how Mastodon unfocuses with Esc.

**Comment by nolanlawson** [comment]
Mastodon doesn't seem to unfocus with the Esc key? Although maybe that's a reasonable workaround: when you press the Esc key when focused on a textbox, it closes the reply box. This is a special behavior we could just do for reply boxes, although I'm not sure how discoverable it is.

**Comment by shinenelson** [comment]
this has been irritating me as well. but my irritation is with all `textarea` fields. One cannot get out of a `textarea` 'trap'. There is the `Tab` way, but that is only a workaround and not a solution. In my humble opinion, the right solution would be to unfocus with the `Esc` key.

I understand that this is kind of broadening the scope of this issue from the title when I say that this affects all `textarea`. I can open a new issue to address all `textarea` fields if necessary.

I'm willing to work on this if given some pointers. I can't seem to figure out where to put a setting that would impact all the `textarea` fields across the application. I've identified the following files within `src/routes/_components` to have `textarea` fields :
- `compose/ComposeInput.html`
- `compose/ComposeMediaItem.html`
- `dialog/components/MediaAltEditor.html`
- `dialog/components/ReportDialog.html`

this is coming from `git grep --files-with-matches textarea`. there are other files that match, but does not have actual `textarea` fields.

should each of the fields be configured individually or can we introduce a global setting that works across the whole application? I'm thinking something like `function acceptShortcutEvent` in `src/routes/_utils/shortcuts.js` but I can't figure out where to put it though.

**Comment by shinenelson** [comment]
I just found #1268 which already adds `Escape` to the global shortcut scope. We'll probably have to generalize that `dismiss()` and move it out of `src/routes/_components/Tooltip.html`. I think I have a starting point. Let me try it out.

**Comment by nolanlawson** [comment]
I think `acceptShortcutEvent` is indeed the part that does special handling for inputs and textareas:

https://github.com/nolanlawson/pinafore/blob/cc8c605828a787f612c2ad60ed5d982395dccd04/src/routes/_utils/shortcuts.js#L173-L186

**Comment by shinenelson** [comment]
damn, my bad. I did say that I would try this out, but I didn't.

I had found `acceptShortcutEvent` earlier and my plan was to generalize the `dismiss()` in `src/routes/_components/Tooltip.html` so that it could go into `src/routes/_utils/shortcuts.js`

I got excited about other parts of the application and lost focus on this. I will get back to this.

**Comment by jcsteh** [comment]
FWIW, with regard to discoverability, I keep trying escape to dismiss accidental invocations of reply and  keep being surprised when it doesn't work. I'm a sample size of 1, so obviously take this with a grain of salt, but I think escape would be fairly intuitive.
